File: examples/minimart-go-nuxt/backend/internal/infrastructure/controller/auth_controller.go
package controller

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"

	"minimart/internal/application/dto"
	"minimart/internal/application/usecase"
	"minimart/internal/domain/entity"
	"minimart/internal/infrastructure/middleware"
)

// AuthController serves POST /auth/login and GET /me.
type AuthController struct {
	auth         *usecase.AuthUsecase
	jwtSecret    string
	cookieSecure bool
}

// NewAuthController returns an AuthController.
func NewAuthController(auth *usecase.AuthUsecase, jwtSecret string, cookieSecure bool) *AuthController {
	return &AuthController{auth: auth, jwtSecret: jwtSecret, cookieSecure: cookieSecure}
}

// Login exchanges credentials for the session cookies.
//
// The response body carries the username and role and nothing else: not the
// password hash, and not the token, which exists only as an HttpOnly cookie.
// Alongside it goes the readable session_hint flag, which carries no authority
// and exists so the front end can know it has a session without asking a
// protected endpoint (see middleware.SetSessionCookies).
func (ctl *AuthController) Login(c *gin.Context) {
	var req dto.LoginRequest
	if err := decodeJSON(c, &req); err != nil {
		respondError(c, err)
		return
	}
	if req.Username == "" || req.Password == "" {
		respondError(c, entity.ErrInvalidRequest)
		return
	}

	user, err := ctl.auth.Authenticate(c.Request.Context(), req.Username, req.Password)
	if err != nil {
		respondError(c, err)
		return
	}

	token, err := middleware.IssueToken(ctl.jwtSecret, user, time.Now())
	if err != nil {
		respondError(c, err)
		return
	}
	middleware.SetSessionCookies(c, token, ctl.cookieSecure)

	c.JSON(http.StatusOK, dto.LoginResponse{Username: user.Username, Role: user.Role})
}

// Me returns the caller's own username, role and points balance.
func (ctl *AuthController) Me(c *gin.Context) {
	userID, err := callerID(c)
	if err != nil {
		respondError(c, err)
		return
	}
	me, err := ctl.auth.Me(c.Request.Context(), userID)
	if err != nil {
		respondError(c, err)
		return
	}
	c.JSON(http.StatusOK, me)
}

File: examples/minimart-java-next/backend/src/main/java/minimart/infrastructure/web/AuthController.java
package minimart.infrastructure.web;

import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import java.time.Instant;
import minimart.application.dto.LoginRequest;
import minimart.application.dto.LoginResponse;
import minimart.application.dto.MeResponse;
import minimart.application.usecase.AuthUsecase;
import minimart.domain.DomainException;
import minimart.domain.ErrorCode;
import minimart.domain.User;
import minimart.infrastructure.security.JwtTokens;
import minimart.infrastructure.security.SessionCookies;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

/** Serves POST /auth/login and GET /me. */
@RestController
@RequestMapping("/api/v1")
public class AuthController {

  private final AuthUsecase auth;
  private final JwtTokens tokens;
  private final SessionCookies cookies;

  /** A controller over {@code auth}, issuing sessions with {@code tokens} and {@code cookies}. */
  public AuthController(AuthUsecase auth, JwtTokens tokens, SessionCookies cookies) {
    this.auth = auth;
    this.tokens = tokens;
    this.cookies = cookies;
  }

  /**
   * Exchanges credentials for the session cookies.
   *
   * <p>The response body carries the username and role and nothing else: not the password hash, and
   * not the token, which exists only as an HttpOnly cookie. Alongside it goes the readable
   * session_hint flag, which carries no authority and exists so the front end can know it has a
   * session without asking a protected endpoint (see {@link SessionCookies}).
   */
  @PostMapping("/auth/login")
  public LoginResponse login(HttpServletRequest http, HttpServletResponse response) {
    LoginRequest request = JsonBody.parse(http, LoginRequest.class);
    if (isEmpty(request.username()) || isEmpty(request.password())) {
      throw new DomainException(ErrorCode.INVALID_REQUEST);
    }

    User user = auth.authenticate(request.username(), request.password());

    cookies.writeSession(response, tokens.issue(user, Instant.now()));
    return new LoginResponse(user.username(), user.role());
  }

  /** Returns the caller's own username, role and points balance. */
  @GetMapping("/me")
  public MeResponse me(Authentication caller) {
    return auth.me(Caller.id(caller));
  }

  private static boolean isEmpty(String value) {
    return value == null || value.isEmpty();
  }
}
